pkg/provisioner/crdprovisioner_test.go

```diff
@@ -0,0 +1,1285 @@
+			expectedError: nil,
+		},
+		{
+			description: "[Failure] Return AlreadyExists error when an AzVolume CRI exists with same volume name but different CreateVolume request parameters",
+			existingAzVolumes: []v1alpha1.AzVolume{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      testDiskName,
+						Namespace: provisioner.namespace,
+					},
+					Spec: v1alpha1.AzVolumeSpec{
+						UnderlyingVolume: testDiskName,
+						VolumeCapability: []v1alpha1.VolumeCapability{
+							{
+								AccessDetails: v1alpha1.VolumeCapabilityAccessDetails{
+									AccessType: v1alpha1.VolumeCapabilityAccessBlock,
+								},
+								AccessMode: v1alpha1.VolumeCapabilityAccessModeSingleNodeWriter,
+							},
+						},
+						Parameters: map[string]string{"parameter": "new params"},
+					},
+					Status: v1alpha1.AzVolumeStatus{
+						Detail: &v1alpha1.AzVolumeStatusDetail{
+							ResponseObject: &v1alpha1.AzVolumeStatusParams{
+								VolumeID:      testDiskURI,
+								CapacityBytes: 2,
+							},
+						},
+					},
+				},
+			},
+			volumeName:           testDiskName,
+			definePrependReactor: false,
+			capacity:             &v1alpha1.CapacityRange{},
+			capabilities: []v1alpha1.VolumeCapability{
+				{
+					AccessDetails: v1alpha1.VolumeCapabilityAccessDetails{
+						AccessType: v1alpha1.VolumeCapabilityAccessMount,
+					},
+					AccessMode: v1alpha1.VolumeCapabilityAccessModeSingleNodeWriter,
+				},
+			},
+			parameters:    make(map[string]string),
+			secrets:       make(map[string]string),
+			contentSource: &v1alpha1.ContentVolumeSource{},
+			topology:      &v1alpha1.TopologyRequirement{},
+			expectedError: status.Errorf(codes.AlreadyExists, "Volume with name (%s) already exists with different specifications", testDiskName),
+		},
+	}
+
+	for _, test := range tests {
+		tt := test
+		t.Run(tt.description, func(t *testing.T) {
+			defer func() { provisioner.azDiskClient = fake.NewSimpleClientset() }()
+
+			if tt.existingAzVolumes != nil {
+				existingList := make([]runtime.Object, len(tt.existingAzVolumes))
+				for itr, azVol := range tt.existingAzVolumes {
+					existingList[itr] = &azVol
+				}
+				provisioner.azDiskClient = fake.NewSimpleClientset(existingList...)
+			}
+
+			if tt.definePrependReactor {
+				// Using the tracker to insert new object or
+				// update the existing object as required
+				tracker := provisioner.azDiskClient.(*fake.Clientset).Tracker()
+
+				/*				provisioner.azDiskClient.(*fake.Clientset).Fake.PrependReactor(
+								"get",
+								"azvolumes",
+								func(action testingClient.Action) (bool, runtime.Object, error) {
+									obj, err := tracker.Get(
+										action.GetResource(),
+										action.GetNamespace(),
+										action.(testingClient.GetAction).GetName())
+
+									if err != nil {
+										return true, nil, err
+									}
+									return true, obj, nil
+								})
+				*/
+				provisioner.azDiskClient.(*fake.Clientset).Fake.PrependReactor(
```

Comment:
Are these prependers necessary? They seem to implement default behavior. I was able to use the simple client to create/update objects w/o needing to define a prepender to handle the action.

Reply:
In this case, I need to overwrite the default create and update to append Status to those CRIs so that the conditionFunctions in the provisioner passes.

Reply:
Oh, yes, I see that now.